File: expense-service/src/main/java/com/splitz/expense/model/Expense.java
package com.splitz.expense.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

@Entity
@Table(name = "expenses")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Expense {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long id;

  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "group_id", nullable = false)
  private Group group;

  @Column(nullable = false)
  private String description;

  @Column(nullable = false, precision = 19, scale = 2)
  private BigDecimal amount;

  @Column(nullable = false, length = 3)
  @Builder.Default
  private String currency = "EUR";

  @Column(name = "paid_by", nullable = false)
  private Long paidBy;

  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "category_id")
  private Category category;

  @Column(name = "expense_date")
  private LocalDate expenseDate;

  @Column(columnDefinition = "TEXT")
  private String notes;

  @Column(name = "receipt_url")
  private String receiptUrl;

  @CreationTimestamp
  @Column(name = "created_at", updatable = false)
  private LocalDateTime createdAt;

  @UpdateTimestamp
  @Column(name = "updated_at")
  private LocalDateTime updatedAt;
}

File: frontend-user/src/types/expense.ts
export type SplitType =
  | "EQUAL"
  | "EXACT"
  | "PERCENTAGE"
  | "SHARES"
  | "ADJUSTMENT";

export interface ExpenseSplit {
  id: number;
  userId: number;
  shareAmount: number;
}

export interface Expense {
  id: number;
  groupId: number;
  description: string;
  amount: number;
  currency: string;
  paidBy: number;
  categoryId?: number;
  expenseDate: string;
  notes?: string;
  receiptUrl?: string;
  lastModifiedBy?: number;
  splits: ExpenseSplit[];
  createdAt: string;
  updatedAt: string;
}

export interface SplitRequest {
  userId: number;
  splitType?: SplitType;
  splitValue?: number;
  shareAmount?: number;
}

export interface CreateExpenseRequest {
  description: string;
  amount: number;
  currency?: string;
  paidBy: number;
  categoryId?: number;
  expenseDate?: string;
  notes?: string;
  receiptUrl?: string;
  splitType: SplitType;
  splits: SplitRequest[];
}

export interface UpdateExpenseRequest {
  description?: string;
  amount?: number;
  currency?: string;
  paidBy?: number;
  categoryId?: number;
  expenseDate?: string;
  notes?: string;
  receiptUrl?: string;
  splitType?: SplitType;
  splits?: SplitRequest[];
}
